src/CoreWCF.Primitives/tests/ServiceKnownTypeTests.cs

```diff
@@ -0,0 +1,223 @@
+                });
+            return factory;
+        }
+
+        internal static TContract CreateChannel<TContract>(System.ServiceModel.ChannelFactory<TContract> factory)
+        {
+            factory.Open();
+            TContract channel = factory.CreateChannel();
+            ((System.ServiceModel.Channels.IChannel)channel).Open();
+            return channel;
+        }
+
+        internal static void CloseChannel<TContract>(System.ServiceModel.ChannelFactory<TContract> factory, TContract channel)
+        {
+            ((System.ServiceModel.Channels.IChannel)channel).Close();
+            factory.Close();
+            TestHelper.CloseServiceModelObjects((System.ServiceModel.Channels.IChannel)channel, factory);
+        }
+
+        [Fact]
+        public static void ServiceKnownTypeException()
+        {
+            var factory = ServiceKnownTypeTests.CreateFactory<IPingServiceEx, PingServiceEx>();
+            var channel = ServiceKnownTypeTests.CreateChannel(factory);
+            Assert.Throws<System.Xml.XmlException>(() =>
+            {
+                BaseMsg msg = channel.Ping("hello");
+            });
+            ServiceKnownTypeTests.CloseChannel(factory, channel);
+        }
+
+        [Fact]
+        public static void ServiceKnownTypeTest1()
+        {
+            var factory = ServiceKnownTypeTests.CreateFactory<IPingService1, PingService1>();
+            var channel = ServiceKnownTypeTests.CreateChannel(factory);
+            BaseMsg msg = channel.Ping("hello");
+            Assert.Equal("hello", msg.msg);
+            ServiceKnownTypeTests.CloseChannel(factory, channel);
+        }
+
+        [Fact]
+        public static void ServiceKnownTypeTest2()
+        {
+            var factory = ServiceKnownTypeTests.CreateFactory<IPingService2, PingService2>();
+            var channel = ServiceKnownTypeTests.CreateChannel(factory);
+            BaseMsg msg = channel.Ping("hello");
+            Assert.Equal("hello", msg.msg);
+            ServiceKnownTypeTests.CloseChannel(factory, channel);
+        }
+
+        [Fact]
+        public static void ServiceKnownTypeTest3()
+        {
+            var factory = ServiceKnownTypeTests.CreateFactory<IPingService3, PingService3>();
+            var channel = ServiceKnownTypeTests.CreateChannel(factory);
+            BaseMsg msg = channel.Ping("hello");
+            Assert.Equal("hello", msg.msg);
+            ServiceKnownTypeTests.CloseChannel(factory, channel);
+        }
+
+        [Fact]
+        public static void ServiceKnownTypeTest4()
+        {
+            var factory = ServiceKnownTypeTests.CreateFactory<IPingService4, PingService4>();
+            var channel = ServiceKnownTypeTests.CreateChannel(factory);
+            BaseMsg msg = channel.Ping("hello");
+            Assert.Equal("hello", msg.msg);
+            ServiceKnownTypeTests.CloseChannel(factory, channel);
+        }
+    }
+
+    internal class PingServiceEx : IPingServiceEx
+    {
```

Comment:
There's a lot of redundancy in these tests. If you declare a base `PingServiceBase` class like this:
```c#
internal class PingServiceBase
{
    public BaseMsg Ping(string msg) => new DerivedMsg { msg = msg };
}
```
Then each of the implementation classes like `PingServiceEx` just need to derive from this class and declare the service interface they are implementing. For example, `PingServiceEx` now becomes this simple one liner:
```c#
internal class PingServiceEx : PingServiceBase, IPingServiceEx { }
```

Reply:
Good suggestion. Implemented.